iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/wal/io/WALInputStream.java

```diff
@@ -190,33 +204,43 @@ private void loadNextSegmentV1() throws IOException {
   }
 
   private void loadNextSegmentV2() throws IOException {
-    headerBuffer.clear();
-    if (channel.read(headerBuffer) != Integer.BYTES + 1) {
-      throw new IOException("Unexpected end of file");
-    }
-    // compressionType originalSize compressedSize
-    headerBuffer.flip();
-    CompressionType compressionType = CompressionType.deserialize(headerBuffer.get());
-    int dataBufferSize = headerBuffer.getInt();
-    if (compressionType != CompressionType.UNCOMPRESSED) {
-      compressedHeader.clear();
-      if (channel.read(compressedHeader) != Integer.BYTES) {
-        throw new IOException("Unexpected end of file");
+    SegmentInfo segmentInfo = getNextSegmentInfo();
+    if (segmentInfo.compressionType != CompressionType.UNCOMPRESSED) {
+      // A compressed segment
+      if (Objects.isNull(dataBuffer)
+          || dataBuffer.capacity() < segmentInfo.uncompressedSize
+          || dataBuffer.capacity() > segmentInfo.uncompressedSize * 2) {
+        dataBuffer = ByteBuffer.allocateDirect(segmentInfo.uncompressedSize);
```

Comment:
If the previous buffer is not null, it should be manually released through `MmapUtil.clean`—the same for the below.